### DISCUSSION TITLE
Feature Request : Duplicate Method

### DISCUSSION THREAD

**Discussion by peddroelm** [dnSpyEx/dnSpy#94] [Ideas]
I want to add logging to a method but only when called from certain functions and not from others .. 

Tried to add an extra optional bool parameter in the method I'm intrested on logging,  error ..cannot find method signature .. 

What worked was duplicating the method by hand (create Method ...) but it can much more time consuming (when it has many parameters) than a simple right click on existing method name and Create Duplicate ...

**Comment by ElektroKill** [comment]
Hi,

Adding an option to duplicate a method should be doable and I completely understand the reasoning behind it. Actually coming up with a user-friendly and intuitive way of implementing it is quite hard. Here are some of my ideas which are of course open to discussion and I'm open to feedback regarding them:

Option 1:
Add a new context menu action for duplicating methods, which upon being executed duplicates the method 1:1 together with the method body and adds it to the declaring type of the method being duplicated. This should be the easiest implementation. However, while being the easiest it has its disadvantages. First of all are method names, having two methods in the same type with the same name and signature can lead to issues with finding the correct overload when editing IL bodies in dnSpy and decompiled code not being valid according to C# rules. Having invalid C# code would lead to issues with the Roslyn-based C# code editor in dnSpy as well as using code generated from the Export Project option. This approach also has the user only having the option to duplicate a method into the original methods declaring type.

Option 2 - an extension of Option 1:
Implement the same context menu action as described in Option 1, but after the user executes the duplicate command, they are prompted to enter a new method name for the duplicated method. This allows the user to easily change the method name when duplicating methods fixing the method naming concern underlined in Option 1.

Option 3 - an extension of Option 2:
Implement the same idea as in Option 2, but extend the dialog to prompt the user to select the type they want the duplicated method to be placed in. This would resolve all issues suggested in Options 1 and 2. However, this option comes with a significant challenge. Cloning methods works well as long as we are within the scope of one .NET module. The second we want to clone a member into a different module we run into problems with importing member references. For now, I think this option is the best as it allows for the most flexibility, and as to the importing problem, we could in theory disallow cross-module duplication of methods unless lots of importing logic could be reused from the dnSpy merge assembly option.

I'm open to other ideas and feedback on this concept :)

**Reply by peddroelm** [reply]
(with my limited understanding ) I Think option 2 and option 3 are different features :) ..   

For option 2 you could auto add a string like "duplicated / clone" at the end of the duplicated method name .. I think the user can simply(is it as simple as Edit Method  .. Alt-Enter OK     ? )  rename it after if it so fancies ..

Could even go crazy and as for number of duplicates ( is there a possible use case that would you need more than one duplicate ?  )  and call them originalmethodname + duplicate01 .. + duplicate0N ? :)
